### PR TITLE
24 benchmark rollup state transition

### LINKED ISSUE #24: Benchmark verifier for zkFold rollup state transition
Objective is to benchmark `rollup` verifier in [Rollups.hs](https://github.com/zkFold/zkfold-cardano/blob/main/src/ZkFold/Cardano/Scripts/Rollup.hs) in order to find how large can the data carried in the redeemer be.

- Redeemer is a tupple with five elements.  Fifth element is a list of numbers (field elements).
- Need to figure out how large can this list of numbers be before the exec budget is reached.

### PR DESCRIPTION
**Pull request by ajuggler**
Benchmarked `rollup` script as a function of *update length*.  See [README](https://github.com/zkFold/zkfold-cardano/blob/24-benchmark-rollup-state-transition/benchs/bench-rollup/README.md) .

Resolves #24 .

### CODE REVIEW

**Comment by ajuggler** [comment]
See summary of results in [README](https://github.com/zkFold/zkfold-cardano/blob/24-benchmark-rollup-state-transition/benchs/bench-rollup/README.md).

File: benchs/bench-rollup/Main.hs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  where e = ""
+
+contextRollup :: ProofBytes -> Int -> ScriptContext
+contextRollup proof n = ScriptContext
+  { scriptContextTxInfo = TxInfo
+    { txInfoInputs = []                     :: [TxInInfo]
+    , txInfoReferenceInputs = []            :: [TxInInfo]
+    , txInfoOutputs = []                    :: [V2.TxOut]
+    , txInfoFee = V2.Lovelace 0
+    , txInfoMint = Value $ unsafeFromList [(dummyCurrencySymbol, unsafeFromList [(dummyTokenName, 1)])]
```
Comment (vlasin) [inline benchs/bench-rollup/Main.hs:64]: The rollup script is meant to be an unspent output validator, not a minting policy.

File: benchs/bench-rollup/Main.hs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+dummyTokenName = TokenName $ toBuiltin (fromString "34ad74db78700c335968ca0898f2953adba88f368efa0541b98897e2668090bd" :: BS.ByteString)
+
+dummyRedeemer :: ProofBytes
+dummyRedeemer = ProofBytes e e e e e e e e e 0 0 0 0 0 0 (F 0)
+  where e = ""
+
+contextRollup :: ProofBytes -> Int -> ScriptContext
+contextRollup proof n = ScriptContext
+  { scriptContextTxInfo = TxInfo
+    { txInfoInputs = []                     :: [TxInInfo]
```
Comment (vlasin) [inline benchs/bench-rollup/Main.hs:60]: Realistically, we need to supply at least two outputs:
- Script output at the rollup address;
- Pub key output of the sequencer node that submits the transaction.

File: benchs/bench-rollup/Main.hs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    , txInfoReferenceInputs = []            :: [TxInInfo]
```
Comment (vlasin) [inline benchs/bench-rollup/Main.hs:61]: One output with the script must be referenced.

File: benchs/bench-rollup/Main.hs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    , txInfoOutputs = []                    :: [V2.TxOut]
```
Comment (vlasin) [inline benchs/bench-rollup/Main.hs:62]: Two outputs must be produced:
- Script output at the rollup address;
- Pub key output with the change paid back to the sequencer node.

File: benchs/bench-rollup/Main.hs (hunk 1 above)
Comment (ajuggler) [reply benchs/bench-rollup/Main.hs]: Done:  cleared `txInfoMint` field.  Also, modified fields `txInfoRedeemers` and `scriptContextScriptInfo` to reflect that it is a spending validator.

File: benchs/bench-rollup/Main.hs (hunk 2 above)
Comment (ajuggler) [reply benchs/bench-rollup/Main.hs]: Done:  provided both outputs.  For the script output, I attached a "unit" datum.

Should we assume that this datum carries info?  If this is the case, what would be a reasonable assumption about its byte size?

File: benchs/bench-rollup/Main.hs (hunk 3 above)
Comment (ajuggler) [reply benchs/bench-rollup/Main.hs]: Done:  a script is now referenced in `txInfoReferenceInputs` field.

File: benchs/bench-rollup/Main.hs (hunk 4 above)
Comment (ajuggler) [reply benchs/bench-rollup/Main.hs]: Done:  added a script output and a pub key output to `txInfoOutputs` field.

**Comment by ajuggler** [comment]
After modifications to `ScriptContext`, exec units increased slightly.  We still arrive at the same conclusion:  "update length" limit is 35.  Can compare `data-analysis-old1.pdf` with the new `data-analysis.pdf` in directory [data-analysis](https://github.com/zkFold/zkfold-cardano/tree/24-benchmark-rollup-state-transition/benchs/bench-rollup/data-analysis) .

**Review by vlasin — APPROVED** [review]

**Pull request merged**